[filebeat][websocket] - Added runtime URL modification support based on state and cursor values

ShourieG opened the pull request:
## Type of change
- Enhancement

## Proposed commit message
Till now it was not possible to modify the Websocket URL dynamically via the CEL program as the Websocket logic exists
outside the scope of the CEL engine instance. To make this action possible a new instance of the CEL engine was added as a one time execution process to prime the URL before the main CEL program executes. This new CEL instance shares the same state hence its possible to modify and use cursor and state values and in turn modify the URL based off of these values.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Beats.
-->

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

- Closes #39858

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

ShourieG commented on x-pack/filebeat/input/websocket/input_test.go:
```diff
@@ -432,7 +529,7 @@ func TestInput(t *testing.T) {
 			}
 
-			ctx, cancel := context.WithTimeout(context.Background(), 1000*time.Second)
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
This was unnecessarily large for some reason, hence adjusted it.

elasticmachine commented:
Pinging @elastic/security-service-integrations (Team:Security-Service Integrations)

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

pierrehilbert commented:
Hey @ShourieG
I added the necessary label to get your PR reviewed by our team.
Could you please make the linter happy by excluding those lines from the linter check? 
As this is not real credential leak, we should make it aware of it.

ShourieG commented on x-pack/filebeat/docs/inputs/input-websocket.asciidoc:
```diff
@@ -153,6 +153,38 @@ program: |
+      : 	
+        state.url+"?since="+ state.initial_start_time
+      )
+	}
```
its properly indented in my local editor, but something is messing up the indentation here.

pierrehilbert commented:
run docs-build

belimawr commented on x-pack/filebeat/input/websocket/input_test.go:
```diff
@@ -399,6 +401,164 @@ var inputTests = []struct {
+			}
+
+			response, err := input{test.time, conf}.initializeInputURL(ctx, state, logp.NewLogger("websocket_input_initializer_test"))
+			if (fmt.Sprint(err) != fmt.Sprint(ctxCancelledError)) && (strings.Split(fmt.Sprint(err), "\n")[0] != fmt.Sprint(test.wantErr)) {
```
[Question]
Why use `fmt.Sprint(err)` instead the `Error()` method?

Also, if you're not using the error value from `ctxCancelledError`, it could just be a string instead of an error.
```suggestion
			if (err.Error() != ctxCancelledError.Error() && (strings.Split(err.Error(), "\n")[0] != test.wantErr.Error()) {
```

ShourieG replied:
This is because it's allows us to avoid contitional checks in every scenario to see if err is nil or not. The Error() method will panic If error is nil, which it will be in many scenarios.

efd6 commented on x-pack/filebeat/input/websocket/input_test.go:
```diff
@@ -399,6 +401,164 @@ var inputTests = []struct {
+
+func TestInputInitializer(t *testing.T) {
+	// tests will ignore context cancelled errors, since they are expected
+	ctxCancelledError := fmt.Errorf("context canceled")
```
```suggestion
```

efd6 commented on x-pack/filebeat/input/websocket/input_test.go:
```diff
@@ -399,6 +401,164 @@ var inputTests = []struct {
+			}
+
+			response, err := input{test.time, conf}.initializeInputURL(ctx, state, logp.NewLogger("websocket_input_initializer_test"))
+			if (fmt.Sprint(err) != fmt.Sprint(ctxCancelledError)) && (strings.Split(fmt.Sprint(err), "\n")[0] != fmt.Sprint(test.wantErr)) {
```
```suggestion
			if !errors.Is(err, context.Canceled) && strings.Split(fmt.Sprint(err), "\n")[0] != fmt.Sprint(test.wantErr) {
```
but I'm not convinced that we need syntax error checking here. This would be akin to ensuring that addition works, these behaviours are already tested in the library.

ShourieG commented:
@efd6, I've addressed all the PR changes and removed the negative syntax tests. @belimawr, your concerns got addressed as the conditions were removed with the latest updates.
